paddle/fluid/platform/device_context.h

```diff
@@ -100,6 +101,7 @@ class CUDADeviceContext : public DeviceContext {
 
   template <typename Callback>
   void RecordEvent(cudaEvent_t ev, Callback callback) {
+    std::lock_guard<std::mutex> guard(mtx_);
     callback();
```

Comment:
How about replacing `std::mutex` with `std::recursive_mutex`? Because, if the `callback()` contains another `RecordEvent()` calling, it will be hang again.

Reply:
This situation should not happen.